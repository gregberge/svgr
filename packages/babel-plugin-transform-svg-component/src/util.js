export const getProps = ({ types: t }, opts) => {
  const props = []

  if (opts.ref) {
    props.push(
      t.objectProperty(
        t.identifier('svgRef'),
        t.identifier('svgRef'),
        false,
        true,
      ),
    )
  }

  if (opts.titleProp) {
    props.push(
      t.objectProperty(
        t.identifier('title'),
        t.identifier('title'),
        false,
        true,
      ),
    )
  }

  if (opts.expandProps) {
    props.push(t.restElement(t.identifier('props')))
  }

  if (props.length === 0) {
    return null
  }

  if (props.length === 1 && opts.expandProps) {
    return t.identifier('props')
  }

  return t.objectPattern(props)
}

export const getImport = ({ types: t }, opts) => {
  const importDeclarations = [
    t.importDeclaration(
      [t.importDefaultSpecifier(t.identifier('React'))],
      t.stringLiteral('react'),
    ),
  ]

  if (opts.native) {
    importDeclarations.push(
      t.importDeclaration(
        [t.importDefaultSpecifier(t.identifier('Svg'))],
        t.stringLiteral('react-native-svg'),
      ),
    )
  }

  return importDeclarations
}

export const getExport = ({ template }, opts) => {
  let result = ''
  let exportName = opts.state.componentName

  if (opts.ref) {
    exportName = 'ForwardRef'
    result += `const ForwardRef = React.forwardRef((props, ref) => <${
      opts.state.componentName
    } svgRef={ref} {...props} />)\n\n`
  }

  if (opts.state.caller && opts.state.caller.previousExport) {
    result += `${opts.state.caller.previousExport}\n`
    result += `export { ${exportName} as ReactComponent }`
    return template.ast(result, {
      plugins: ['jsx'],
    })
  }

  result += `export default ${exportName}`
  return template.ast(result, {
    plugins: ['jsx'],
  })
}
