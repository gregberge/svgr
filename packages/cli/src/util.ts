/* eslint-disable no-console */
import { promises as fs } from 'fs'
import { red } from 'chalk'
import { transform, Config, State } from '@svgr/core'
import svgo from '@svgr/plugin-svgo'
import jsx from '@svgr/plugin-jsx'
import prettier from '@svgr/plugin-prettier'
// @ts-ignore
import camelCase from 'camelcase'
// @ts-ignore
import dashify from 'dashify'

export function transformFilename(
  filename: string,
  filenameCase: string,
): string {
  switch (filenameCase) {
    case 'kebab':
      return dashify(filename.replace(/_/g, '-'), { condense: true })
    case 'camel':
      return camelCase(filename)
    case 'pascal':
      return camelCase(filename, { pascalCase: true })
    default:
      throw new Error(`Unknown --filename-case ${filenameCase}`)
  }
}

export const convert = (
  code: string,
  config: Config,
  state: Partial<State>,
): string => {
  return transform.sync(code, config, {
    ...state,
    caller: {
      name: '@svgr/cli',
      defaultPlugins: [svgo, jsx, prettier],
    },
  })
}

export const convertFile = async (
  filePath: string,
  config: Config = {},
): Promise<string> => {
  const code = await fs.readFile(filePath, 'utf-8')
  return convert(code, config, { filePath })
}

export const exitError = (error: string): never => {
  console.error(red(error))
  process.exit(1)
}

export const politeWrite = (data: string, silent?: boolean): void => {
  if (!silent) {
    process.stdout.write(data)
  }
}

export const formatExportName = (name: string): string => {
  if (/[-]/g.test(name) && /^\d/.test(name)) {
    return `Svg${camelCase(name, { pascalCase: true })}`
  }

  if (/^\d/.test(name)) {
    return `Svg${name}`
  }

  return camelCase(name, { pascalCase: true })
}
