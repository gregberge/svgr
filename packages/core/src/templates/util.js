export const getProps = config => {
  const props = []
  if (config.ref) props.push('svgRef')
  if (config.titleProp) props.push('title')
  if (config.expandProps) props.push('...props')

  if (props.length === 0) return '()'
  if (props.length === 1 && config.expandProps) return 'props'

  return `({ ${props.join(', ')} })`
}
