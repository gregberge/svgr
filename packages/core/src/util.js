import path from 'path'
import { upperFirst, camelCase } from 'lodash'

export const pascalCase = str => upperFirst(camelCase(str))

export function getComponentName(state) {
  if (!state.filePath) return 'SvgComponent'
  const componentName = pascalCase(path.parse(state.filePath).name)
  if (Number.isNaN(parseInt(componentName[0], 10))) return componentName
  return `Svg${componentName}`
}

export function expandState(state) {
  return {
    componentName: state.componentName || getComponentName(state),
    ...state,
  }
}
