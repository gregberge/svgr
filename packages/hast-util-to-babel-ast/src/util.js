/**
 * Determines if the specified string consists entirely of numeric characters.
 *
 * @param {*} [value]
 * @returns {boolean}
 */
export function isNumeric(value) {
  return !Number.isNaN(value - parseFloat(value))
}

/**
 * Convert a hyphenated string to camelCase.
 *
 * @param {string} string
 * @returns {string}
 */
export function hyphenToCamelCase(string) {
  return string.replace(/-(.)/g, (match, chr) => chr.toUpperCase())
}

/**
 * Trim the specified substring off the string. If the string does not end
 * with the specified substring, this is a no-op.
 *
 * @param {string} haystack String to search in
 * @param {string} needle   String to search for
 * @return {string}
 */
export function trimEnd(haystack, needle) {
  return haystack.endsWith(needle)
    ? haystack.slice(0, -needle.length)
    : haystack
}
