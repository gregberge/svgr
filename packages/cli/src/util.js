/* eslint-disable no-console */
import fs from 'fs'
import chalk from 'chalk'
import util from 'util'
import convert, { resolveConfig } from '@svgr/core'

export const readFile = util.promisify(fs.readFile)
export const stat = util.promisify(fs.stat)

export async function convertFile(filePath, { config, ...options } = {}) {
  const code = await readFile(filePath, 'utf-8')
  const rcConfig = await resolveConfig(filePath, config)
  return convert(code, { ...rcConfig, ...options }, { filePath })
}

export function exitError(error) {
  console.error(chalk.red(error))
  process.exit(1)
}
