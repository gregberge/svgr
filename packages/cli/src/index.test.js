import fs from 'fs'
import path from 'path'
import childProcess from 'child_process'
import util from 'util'
import del from 'del'

const readdir = util.promisify(fs.readdir)
const exec = util.promisify(childProcess.exec)

const svgr = path.join(__dirname, 'index.js')
const babelNode = path.join(__dirname, '../../../node_modules/.bin/babel-node')

describe('cli', () => {
  const cli = async args => {
    const { stdout } = await exec(`${babelNode} -- ${svgr} ${args}`)
    return stdout
  }

  it('should work with a simple file', async () => {
    const result = await cli('__fixtures__/simple/file.svg')
    expect(result).toMatchSnapshot()
  }, 10000)

  it('should not work with a directory without --out-dir option', async () => {
    expect.assertions(1)
    try {
      await cli('__fixtures__/nesting')
    } catch (error) {
      expect(error.message).toMatch(
        'Directory are not supported without `--out-dir` option instead',
      )
    }
  }, 10000)

  it('should not work with several files without destination', async () => {
    expect.assertions(1)
    try {
      await cli('__fixtures__/simple/file.svg __fixtures__/nesting/one.svg')
    } catch (error) {
      expect(error.message).toMatch(
        'Please specify only one filename or use `--out-dir` option',
      )
    }
  }, 10000)

  it('should work with stdin', async () => {
    const result = await cli('< __fixtures__/simple/file.svg')
    expect(result).toMatchSnapshot()
  }, 10000)

  it('should transform a whole directory', async () => {
    await cli('--out-dir __fixtures_build__/whole __fixtures__')
  }, 10000)

  it('should support --prettier-config as json', async () => {
    const result = await cli(
      `--prettier-config '{"tabWidth": 5}' __fixtures__/simple/file.svg`,
    )
    expect(result).toMatchSnapshot()
  }, 10000)

  it('should support --prettier-config as file', async () => {
    const result = await cli(
      `--prettier-config __fixtures__/withPrettierRc/.prettierrc __fixtures__/simple/file.svg`,
    )
    expect(result).toMatchSnapshot()
  }, 10000)

  it('should support --svgo-config as json', async () => {
    const result = await cli(
      `--svgo-config '{"plugins": [{"removeTitle": false}]}' __fixtures__/simple/file.svg`,
    )
    expect(result).toMatchSnapshot()
  }, 10000)

  it('should support --svgo-config as file', async () => {
    const result = await cli(
      `--svgo-config __fixtures__/withSvgoYml/.svgo.json __fixtures__/simple/file.svg`,
    )
    expect(result).toMatchSnapshot()
  }, 10000)

  it.each([
    ['--no-dimensions'],
    ['--expand-props none'],
    ['--expand-props start'],
    ['--icon'],
    ['--native'],
    ['--native --icon'],
    ['--native --expand-props none'],
    ['--native --ref'],
    ['--ref'],
    ['--replace-attr-values "#063855=currentColor"'],
    [`--svg-props "hidden={true}"`],
    ['--no-svgo'],
    ['--no-prettier'],
    ['--title-prop'],
  ])(
    'should support various args',
    async args => {
      const result = await cli(`${args} __fixtures__/simple/file.svg`)
      expect(result).toMatchSnapshot(args)
    },
    10000,
  )

  it.each([
    [0, ''],
    [1, '--filename-case=camel'],
    [2, '--filename-case=pascal'],
    [3, '--filename-case=kebab'],
  ])(
    'should support different filename cases with directory output',
    async (index, args) => {
      const inDir = '__fixtures__/cased'
      const outDir = `__fixtures_build__/filename-case-${index}`
      await del(outDir)
      await cli(`${args} ${inDir} --out-dir=${outDir}`)
      expect(await readdir(outDir)).toMatchSnapshot(args)
    },
    10000,
  )

  it('should support custom file extension', async () => {
    const inDir = '__fixtures__/simple'
    const outDir = '__fixtures_build__/ext'
    await del(outDir)
    await cli(`--ext=ts ${inDir} --out-dir=${outDir}`)
    expect(await readdir(outDir)).toMatchSnapshot()
  }, 10000)

  it('should not override config with cli defaults', async () => {
    const result = await cli(
      '__fixtures__/simple/file.svg --config-file=__fixtures__/overrides.config.js',
    )
    expect(result).toMatchSnapshot()
  }, 10000)
})
