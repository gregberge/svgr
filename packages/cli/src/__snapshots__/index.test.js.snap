// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`cli should not override config with cli defaults 1`] = `
"import React from 'react'

const File = () => <svg viewBox=\\"0 0 48 1\\" version={1.1} xmlnsXlink=\\"http://www.w3.org/1999/xlink\\">
  <title>
    Rectangle 5
  </title>
  <desc>
    Created with Sketch.
  </desc>
  <defs />
  <g id=\\"Page-1\\" stroke=\\"none\\" strokeWidth={1} fill=\\"none\\" fillRule=\\"evenodd\\">
    <g id=\\"19-Separator\\" transform=\\"translate(-129.000000, -156.000000)\\" fill=\\"#063855\\">
      <g id=\\"Controls/Settings\\" transform=\\"translate(80.000000, 0.000000)\\">
        <g id=\\"Content\\" transform=\\"translate(0.000000, 64.000000)\\">
          <g id=\\"Group\\" transform=\\"translate(24.000000, 56.000000)\\">
            <g id=\\"Group-2\\">
              <rect id=\\"Rectangle-5\\" x={25} y={36} width={48} height={1} />
            </g>
          </g>
        </g>
      </g>
    </g>
  </g>
</svg>


export default File
"
`;

exports[`cli should support --prettier-config as file 1`] = `
"import React from 'react'

const File = props => (
     <svg width={48} height={1} {...props}>
          <path d=\\"M0 0h48v1H0z\\" fill=\\"#063855\\" fillRule=\\"evenodd\\" />
     </svg>
)

export default File

"
`;

exports[`cli should support --prettier-config as json 1`] = `
"import React from 'react'

const File = props => (
     <svg width={48} height={1} {...props}>
          <path d=\\"M0 0h48v1H0z\\" fill=\\"#063855\\" fillRule=\\"evenodd\\" />
     </svg>
)

export default File

"
`;

exports[`cli should support --svgo-config as file 1`] = `
"import React from 'react'

const File = props => (
  <svg width={48} height={1} {...props}>
    <title>Rectangle 5</title>
    <path d=\\"M0 0h48v1H0z\\" fill=\\"#063855\\" fillRule=\\"evenodd\\" />
  </svg>
)

export default File

"
`;

exports[`cli should support --svgo-config as json 1`] = `
"import React from 'react'

const File = props => (
  <svg width={48} height={1} {...props}>
    <title>Rectangle 5</title>
    <path d=\\"M0 0h48v1H0z\\" fill=\\"#063855\\" fillRule=\\"evenodd\\" />
  </svg>
)

export default File

"
`;

exports[`cli should support custom file extension 1`] = `
Array [
  "File.ts",
]
`;

exports[`cli should support different filename cases with directory output 1`] = `
Array [
  "CamelCase.js",
  "KebabCase.js",
  "MultipleDashes.js",
  "PascalCase.js",
]
`;

exports[`cli should support different filename cases with directory output: --filename-case=camel 1`] = `
Array [
  "camelCase.js",
  "kebabCase.js",
  "multipleDashes.js",
  "pascalCase.js",
]
`;

exports[`cli should support different filename cases with directory output: --filename-case=kebab 1`] = `
Array [
  "camel-case.js",
  "kebab-case.js",
  "multiple-dashes.js",
  "pascal-case.js",
]
`;

exports[`cli should support different filename cases with directory output: --filename-case=pascal 1`] = `
Array [
  "CamelCase.js",
  "KebabCase.js",
  "MultipleDashes.js",
  "PascalCase.js",
]
`;

exports[`cli should support various args: --icon 1`] = `
"import React from 'react'

const File = props => (
  <svg viewBox=\\"0 0 48 1\\" width=\\"1em\\" height=\\"1em\\" {...props}>
    <path d=\\"M0 0h48v1H0z\\" fill=\\"#063855\\" fillRule=\\"evenodd\\" />
  </svg>
)

export default File

"
`;

exports[`cli should support various args: --native --icon 1`] = `
"import React from 'react'
import Svg, { Path } from 'react-native-svg'
const File = props => (
  <Svg viewBox=\\"0 0 48 1\\" width=\\"1em\\" height=\\"1em\\" {...props}>
    <Path d=\\"M0 0h48v1H0z\\" fill=\\"#063855\\" fillRule=\\"evenodd\\" />
  </Svg>
)

export default File

"
`;

exports[`cli should support various args: --native --no-expand-props 1`] = `
"import React from 'react'
import Svg, { Path } from 'react-native-svg'
const File = () => (
  <Svg width={48} height={1}>
    <Path d=\\"M0 0h48v1H0z\\" fill=\\"#063855\\" fillRule=\\"evenodd\\" />
  </Svg>
)

export default File

"
`;

exports[`cli should support various args: --native --ref 1`] = `
"import React from 'react'
import Svg, { Path } from 'react-native-svg'
const File = ({ svgRef, ...props }) => (
  <Svg width={48} height={1} ref={svgRef} {...props}>
    <Path d=\\"M0 0h48v1H0z\\" fill=\\"#063855\\" fillRule=\\"evenodd\\" />
  </Svg>
)

export default File

"
`;

exports[`cli should support various args: --native 1`] = `
"import React from 'react'
import Svg, { Path } from 'react-native-svg'
const File = props => (
  <Svg width={48} height={1} {...props}>
    <Path d=\\"M0 0h48v1H0z\\" fill=\\"#063855\\" fillRule=\\"evenodd\\" />
  </Svg>
)

export default File

"
`;

exports[`cli should support various args: --no-dimensions 1`] = `
"import React from 'react'

const File = props => (
  <svg {...props}>
    <path d=\\"M0 0h48v1H0z\\" fill=\\"#063855\\" fillRule=\\"evenodd\\" />
  </svg>
)

export default File

"
`;

exports[`cli should support various args: --no-expand-props 1`] = `
"import React from 'react'

const File = () => (
  <svg width={48} height={1}>
    <path d=\\"M0 0h48v1H0z\\" fill=\\"#063855\\" fillRule=\\"evenodd\\" />
  </svg>
)

export default File

"
`;

exports[`cli should support various args: --no-prettier 1`] = `
"import React from 'react'

const File = props => <svg width={48} height={1} {...props}>
  <path d=\\"M0 0h48v1H0z\\" fill=\\"#063855\\" fillRule=\\"evenodd\\" />
</svg>


export default File
"
`;

exports[`cli should support various args: --no-svgo 1`] = `
"import React from 'react'

const File = props => (
  <svg
    width=\\"48px\\"
    height=\\"1px\\"
    viewBox=\\"0 0 48 1\\"
    version={1.1}
    xmlnsXlink=\\"http://www.w3.org/1999/xlink\\"
    {...props}
  >
    <title>Rectangle 5</title>
    <desc>Created with Sketch.</desc>
    <defs />
    <g id=\\"Page-1\\" stroke=\\"none\\" strokeWidth={1} fill=\\"none\\" fillRule=\\"evenodd\\">
      <g
        id=\\"19-Separator\\"
        transform=\\"translate(-129.000000, -156.000000)\\"
        fill=\\"#063855\\"
      >
        <g id=\\"Controls/Settings\\" transform=\\"translate(80.000000, 0.000000)\\">
          <g id=\\"Content\\" transform=\\"translate(0.000000, 64.000000)\\">
            <g id=\\"Group\\" transform=\\"translate(24.000000, 56.000000)\\">
              <g id=\\"Group-2\\">
                <rect id=\\"Rectangle-5\\" x={25} y={36} width={48} height={1} />
              </g>
            </g>
          </g>
        </g>
      </g>
    </g>
  </svg>
)

export default File

"
`;

exports[`cli should support various args: --ref 1`] = `
"import React from 'react'

const File = ({ svgRef, ...props }) => (
  <svg width={48} height={1} ref={svgRef} {...props}>
    <path d=\\"M0 0h48v1H0z\\" fill=\\"#063855\\" fillRule=\\"evenodd\\" />
  </svg>
)

export default File

"
`;

exports[`cli should support various args: --replace-attr-values "#063855=currentColor" 1`] = `
"import React from 'react'

const File = props => (
  <svg width={48} height={1} {...props}>
    <path d=\\"M0 0h48v1H0z\\" fill=\\"currentColor\\" fillRule=\\"evenodd\\" />
  </svg>
)

export default File

"
`;

exports[`cli should support various args: --svg-attributes "focusable=false" 1`] = `
"import React from 'react'

const File = props => (
  <svg focusable=\\"false\\" width={48} height={1} {...props}>
    <path d=\\"M0 0h48v1H0z\\" fill=\\"#063855\\" fillRule=\\"evenodd\\" />
  </svg>
)

export default File

"
`;

exports[`cli should support various args: --svg-props "hidden={true}" 1`] = `
"import React from 'react'

const File = props => (
  <svg hidden={true} width={48} height={1} {...props}>
    <path d=\\"M0 0h48v1H0z\\" fill=\\"#063855\\" fillRule=\\"evenodd\\" />
  </svg>
)

export default File

"
`;

exports[`cli should support various args: --title-prop 1`] = `
"import React from 'react'

const File = ({ title, ...props }) => (
  <svg width={48} height={1} {...props}>
    <title>{title}</title>
    <path d=\\"M0 0h48v1H0z\\" fill=\\"#063855\\" fillRule=\\"evenodd\\" />
  </svg>
)

export default File

"
`;

exports[`cli should work with a simple file 1`] = `
"import React from 'react'

const File = props => (
  <svg width={48} height={1} {...props}>
    <path d=\\"M0 0h48v1H0z\\" fill=\\"#063855\\" fillRule=\\"evenodd\\" />
  </svg>
)

export default File

"
`;

exports[`cli should work with stdin 1`] = `
"import React from 'react'

const SvgComponent = props => (
  <svg width={48} height={1} {...props}>
    <path d=\\"M0 0h48v1H0z\\" fill=\\"#063855\\" fillRule=\\"evenodd\\" />
  </svg>
)

export default SvgComponent

"
`;
